src/checked_transaction.rs

```diff
@@ -92,6 +104,77 @@ impl CheckedTransaction {
         self.checked_signatures
     }
 
+    pub const fn metadata(&self) -> Option<&Metadata> {
+        self.transaction.metadata()
+    }
+
+    /// Update change and variable outputs.
+    ///
+    /// `revert` will signal if the execution was reverted. It will refund the unused gas cost to
+    /// the base asset and reset output changes to their initial balances.
+    ///
+    /// `remaining_gas` expects the raw content of `$ggas`
+    ///
+    /// `balances` will contain the current state of the free balances
+    pub fn update_outputs<I>(
+        &mut self,
+        params: &ConsensusParameters,
+        revert: bool,
+        remaining_gas: Word,
+        balances: &I,
+    ) -> Result<(), ValidationError>
+    where
+        I: for<'a> Index<&'a AssetId, Output = Word>,
+    {
+        let factor = params.gas_price_factor as f64;
+        let gas_refund = self
+            .transaction()
+            .gas_price()
+            .checked_mul(remaining_gas)
+            .ok_or(ValidationError::ArithmeticOverflow)? as f64;
+
+        let gas_refund = (gas_refund / factor).floor() as Word;
```

Comment:
We should **never** use f64 when dealing with accounting units like gas or asset values. 

Here's how I think we should safely do this, which is similar to how to implemented the gas limit deduction for the initial free balances:

```suggestion
        let factor = params.gas_price_factor as f64;
        let unfactored_refund = (self.transaction().gas_price() as u128)
            .checked_mul(remaining_gas as u128)
            .ok_or(ValidationError::ArithmeticOverflow)?;
        let gas_refund: Word = div_floor(unfactored_refund, params.gas_price_factor as u128)
            .try_into()
            .map_err(|_| ValidationError::ArithmeticOverflow)?;
```

Reply:
Good catch there!

I just remembered we also included the tx fee structure that should totally encapsulate this behavior so we don't have to repeat this logic somewhere else (therefore, this mistake won't happen again)

https://github.com/FuelLabs/fuel-tx/pull/162/commits/6a22b7b6c66dc9e00cd20cffeb69fb0f0e42ad10